### PR TITLE
[Tooling] Use `buildkite_pipeline_upload` action to inline Buildkite release pipelines

### PR DESCRIPTION
**Pull request by iangmaia**
## Description

Android counterpart: https://github.com/woocommerce/woocommerce-android/pull/12839
Related: pdccJN-xR-p2

This PR uses the `release-toolkit` action `buildkite_pipeline_upload` implemented on https://github.com/wordpress-mobile/release-toolkit/pull/597.

Our release pipelines (such as `complete-code-freeze`, `finalize-release`, etc) usually require to start a separate pipeline to make a release build.
The changes in this PR will result in add the steps of a release pipeline to the currently running build, without starting a new build altogether.

I've also added a trigger to run Danger whenever a PR moves to the "Ready For Review" state, as previously we relied on this behavior from Buildkite, but now it has changed.

## Testing information
I've added (and later reverted) a test commit 1b7d00fe78e0f23d78fdeae40fbfbcd9f0692ee1 to test the `complete_code_freeze` lane.

This is a test build created using the Buildkite Web UI directly with `PIPELINE=release-pipelines/complete-code-freeze.yml`:
- https://buildkite.com/automattic/woocommerce-ios/builds/25608

This PR can be tested in a similar way to the test above: creating a branch off this one, slightly modifying a release lane to prevent side effects and running the build directly on Buildkite.

### CODE REVIEW

File: fastlane/Fastfile [hunk 1]
```diff
@@ -1324,16 +1328,24 @@ end
...
 def trigger_buildkite_release_build(branch:, beta:)
-  build_url = buildkite_trigger_build(
-    buildkite_organization: 'automattic',
-    buildkite_pipeline: 'woocommerce-ios',
-    branch: branch,
-    environment: { BUILDKITE_BETA_RELEASE: beta },
-    pipeline_file: 'release-builds.yml'
-  )
+    # If we're running on CI, we can directly start the release pipeline jobs within the same build
+    if is_ci
```
Comment (dangermattic) [inline fastlane/Fastfile:1332]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-ios/blob/0f223b96ead709c307084ad27ac33338980a54b6/fastlane/Fastfile#L1332"></span>Layout/IndentationWidth: Use 2 (not 4) spaces for indentation.</td>
    </tr>
  </tbody>
</table>

File: fastlane/Fastfile [hunk 2]
```diff
@@ -1324,16 +1328,24 @@ end
...
+      buildkite_pipeline_upload(
+        pipeline_file: BUILDKITE_RELEASE_PIPELINE_FILE_PATH,
+        env_file: BUILDKITE_SHARED_PIPELINE_VARS_FILE_PATH,
+        environment: { BUILDKITE_BETA_RELEASE: beta },
```
Comment (dangermattic) [inline fastlane/Fastfile:1336]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-ios/blob/0f223b96ead709c307084ad27ac33338980a54b6/fastlane/Fastfile#L1336"></span>Style/TrailingCommaInArguments: Avoid comma after the last parameter of a method call.</td>
    </tr>
  </tbody>
</table>

File: fastlane/Fastfile [hunk 3]
```diff
@@ -1324,16 +1328,24 @@ end
...
+      )
+    else
+      build_url = buildkite_trigger_build(
+        buildkite_organization: 'automattic',
+        buildkite_pipeline: 'woocommerce-ios',
+        branch: branch,
+        pipeline_file: BUILDKITE_RELEASE_PIPELINE,
+        environment: { BUILDKITE_BETA_RELEASE: beta },
```
Comment (dangermattic) [inline fastlane/Fastfile:1344]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-ios/blob/0f223b96ead709c307084ad27ac33338980a54b6/fastlane/Fastfile#L1344"></span>Style/TrailingCommaInArguments: Avoid comma after the last parameter of a method call.</td>
    </tr>
  </tbody>
</table>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/230">21.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: .github/workflows/run-danger.yml [hunk 4]
```diff
@@ -2,12 +2,12 @@ name: ☢️ Trigger Danger On Buildkite
...
 on:
   pull_request:
-    types: [labeled, unlabeled, milestoned, demilestoned]
+    types: [labeled, unlabeled, milestoned, demilestoned, ready_for_review]
 
 jobs:
   dangermattic:
     if: ${{ (github.event.pull_request.draft == false) }}
-    uses: Automattic/dangermattic/.github/workflows/reusable-retry-buildkite-step-on-events.yml@v1.1.0
+    uses: Automattic/dangermattic/.github/workflows/reusable-retry-buildkite-step-on-events.yml@v1.1.2.1
```
Comment (mokagio) [inline .github/workflows/run-danger.yml]: In the PR description, you mention this change as "I also added..." which makes me think this is orthogonal to the `buildkite_pipeline_upload` right? 

Asking because I'm doing something similar in Simplenote iOS.

File: fastlane/Fastfile [hunk 5]
```diff
@@ -1324,16 +1326,27 @@ end
...
 def trigger_buildkite_release_build(branch:, beta:)
-  build_url = buildkite_trigger_build(
-    buildkite_organization: 'automattic',
-    buildkite_pipeline: 'woocommerce-ios',
-    branch: branch,
-    environment: { BUILDKITE_BETA_RELEASE: beta },
-    pipeline_file: 'release-builds.yml'
-  )
+  beta = beta.to_s
+  environment = { 'BUILDKITE_BETA_RELEASE' => beta }
```
Comment (mokagio) [inline fastlane/Fastfile]: I'm surprised we need to use the `BUILDKITE_` prefix 🤔

**Review by mokagio — COMMENTED** [review]
@iangmaia in implementing this in Simplenote iOS, I run into the issue @AliSoftware mentioned somewhere recently but cannot find at the moment of the Git branch.

See:

- https://github.com/Automattic/simplenote-ios/pull/1673
- https://github.com/Automattic/simplenote-ios/pull/1675
- https://github.com/Automattic/simplenote-ios/pull/1677

TL;DR – The new beta build lane bumped the version, pushed, then uploaded the release build pipeline. But the release build pipeline run on the commit that _started_ the new beta build, which was before the version bump. I tried a couple of things, and settled for calling the checkout release branch command in CI as part of the `release-build.yml` pipeline commands.

Is the approach I described needed here, too?

What do you think of forcing `release-build.yml` to checkout the release branch? I'm of two minds... I don't like that it constrains on which run the build should run. On the other hand, in practice it always run on `release/*` anyway, so might as well keep things simpler in the rest of the automation...

File: .github/workflows/run-danger.yml (hunk 4 above)
Comment (iangmaia) [reply .github/workflows/run-danger.yml]: Yep, the GHA changes are unrelated to the secondary pipeline upload.

The GHA trigger on `Ready For Review` is required since we're not running the full build again on Buildkite once a PR moves to Ready for Review.

File: fastlane/Fastfile (hunk 5 above)
Comment (iangmaia) [reply fastlane/Fastfile]: ~Yeah, we need this to be propagated to the macOS VM, so if I understand it correctly we still need it.~
Oh, this actually isn't true -- it's used solely in the Buildkite `.yml` pre-processing it seems? I think we can change it. I'll try.

Comment (AliSoftware) [reply fastlane/Fastfile]: > so if I understand it correctly we still need it.

Not if we adopt the alternative approach that we've been adopting in all other repos (resolve env var in `.yml` pass resolved value as parameter to invoked `.sh`), to avoid using env vars that are supposed to be reserved for Buildkite internals only 😉 

See https://github.com/Automattic/hostmgr/issues/110 for more details (and the "Workaround in the meantime" section)

**Comment by iangmaia** [comment]
@mokagio:
> @iangmaia in implementing this in Simplenote iOS, I run into the issue @AliSoftware mentioned somewhere recently but cannot find at the moment of the Git branch.
> 
> TL;DR – The new beta build lane bumped the version, pushed, then uploaded the release build pipeline. But the release build pipeline run on the commit that _started_ the new beta build, which was before the version bump. I tried a couple of things, and settled for calling the checkout release branch command in CI as part of the `release-build.yml` pipeline commands.
> 
> Is the approach I described needed here, too?
> 

Oof, I haven't seen this before but I imagine this will be needed here, too 😞 thanks for letting me know.

> What do you think of forcing `release-build.yml` to checkout the release branch? I'm of two minds... I don't like that it constrains on which run the build should run. On the other hand, in practice it always run on `release/*` anyway, so might as well keep things simpler in the rest of the automation...

I think that makes sense, yes. I thought hotfixes could be an exception (e.g. `hotfix/*`) but I think everything is following the `release/*` convention, so that makes sense to me 👍 I'll update my pending PRs with the same change.

**Comment by iangmaia** [comment]
> > What do you think of forcing `release-build.yml` to checkout the release branch? I'm of two minds... I don't like that it constrains on which run the build should run. On the other hand, in practice it always run on `release/*` anyway, so might as well keep things simpler in the rest of the automation...
> 
> I think that makes sense, yes. I thought hotfixes could be an exception (e.g. `hotfix/*`) but I think everything is following the `release/*` convention, so that makes sense to me 👍 I'll update my pending PRs with the same change.

@mokagio @AliSoftware update: 1722ceb1650b503cdf6ca230db53e2d9968903a3

File: fastlane/Fastfile (hunk 5 above)
Comment (iangmaia) [reply fastlane/Fastfile]: Yeah, I later realized it's already the case (so not sure why it ever needed the `BUILDKITE_` prefix) and updated the comment. It's resolved and used only in the `.yml`, so I updated it on 1722ceb1650b503cdf6ca230db53e2d9968903a3.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by mokagio — APPROVED** [review]
